### PR TITLE
Proposal: refactor code to typescript

### PR DESCRIPTION
**Pull request by sarahsporck**
Working with flow is cumbersome and exhausting 🙈 , so why not use typescript 🔥

### CODE REVIEW

File: src/Headroom.tsx [hunk 1]
```diff
@@ -110,8 +108,8 @@ class Headroom extends React.PureComponent<PropsType, StateType> {
     }
     if (window.pageYOffset !== undefined) {
       return window.pageYOffset
-    } else if (window.scrollTop !== undefined) {
-      return window.scrollTop
+    } else if (window.scrollY !== undefined) {
+      return window.scrollY
```
Comment (sarahsporck) [inline src/Headroom.tsx:112]: This is the only place I actually had to change the content of the code. Apparently window does not have the property `scrollTop` so I guessed `scrollY` would be the equivalent.

**Comment by michael-markl** [comment]
Thanks for the change. I'll have a look at the PR hopefully on the weekend.

File: package.json [hunk 2]
```diff
@@ -31,23 +31,21 @@
...
     "test:update": "jest --config jest.config.json -u",
     "lint": "npm run eslint && npm run stylelint",
     "lint:fix": "npm run eslint --fix && npm run stylelint",
     "eslint": "eslint .",
-    "stylelint": "stylelint './src/**/*.js'",
-    "flow:check-now": "flow check",
-    "flow:install-types": "flow-typed install --skip",
+    "stylelint": "stylelint './src/**/*.{js,ts,tsx}'",
+    "ts:build": "tsc --build",
     "prepublishOnly": "npm run build && npm run build:demo && npm run test && npm run lint && npm run flow:check-now"
```
Comment (michael-markl) [inline package.json:49]: This needs to be adjusted as well.

File: package.json [hunk 3]
```diff
@@ -31,23 +31,21 @@
...
     "test:watch": "jest --config jest.config.json --watchAll",
     "test:update": "jest --config jest.config.json -u",
     "lint": "npm run eslint && npm run stylelint",
     "lint:fix": "npm run eslint --fix && npm run stylelint",
     "eslint": "eslint .",
-    "stylelint": "stylelint './src/**/*.js'",
-    "flow:check-now": "flow check",
-    "flow:install-types": "flow-typed install --skip",
+    "stylelint": "stylelint './src/**/*.{js,ts,tsx}'",
+    "ts:build": "tsc --build",
```
Comment (michael-markl) [inline package.json:48]: What's the difference between "build" and "ts:build"? Do we need both?

File: jest.config.json [hunk 4]
```diff
@@ -4,19 +4,16 @@
...
+    "<rootDir>/setupTest.ts"
   ],
   "globals": {
     "__DEV__": false
   },
-  "collectCoverageFrom": [
-    "**/*.{js,jsx}"
+  "transform": {
+    "^.+\\.(t|j)sx?$": "@swc/jest"
+  },
```
Comment (michael-markl) [inline jest.config.json:14]: What does that do? Should we use babel and swc in the same repo? (I have no experience with swc yet, but am happy with switching to it -- but probably we should replace babel everywhere then?)

**Review by michael-markl — COMMENTED** [review]
@sarahsporck Sorry for the extremely slow review... 😿

 I *am* convinced, that Flow is not the future anymore :)

There's a movement in using typescript only as part of jsdoc annotations (see Svelte.js as one big proponent of this technique) which I am quite interested in. I thought, maybe react-sticky-headroom would be a nice experiment to try it out. 
The main benefits are
* no need for emitting source-map files (which I think we didn't do anyways :D )
* no need for transpilation
* users of the lib can directly jump to the actual source (instead of to type definitions)

making the whole stack (in theory) less complex.

I tried to use it, but at some point, I always get tired/bored by the complexity of the node / typescript / ES module resolution stuff; I haven't found enough time/patience to figure it out.

So I figured now, that we can at least start with your PR migrating to "regular" typescript. I would even say, that we drop flow types completely (and make a breaking major release). I guess, we won't check whether the (generated) types actually work as expected (plus: one less dependency).

File: jest.config.json (hunk 4 above)
Comment (sarahsporck) [reply jest.config.json:14]: If you want to use jest for typescipt files you need to use a proper transformer. I'm really not convinced by babel-jest, and ts-jest, as they both had their issues in the past. Also swc is about twice as fast, as it is implemented in rust :).

I decided to strip babel completely and replace it with babel.

File: package.json (hunk 3 above)
Comment (sarahsporck) [reply package.json]: I renamed ts:build with `ts:check` to make clear that this just checks the typescript types.

File: src/Headroom.tsx (hunk 1 above)
Comment (sarahsporck) [reply src/Headroom.tsx]: You mean to remove the complete if/else if/else block and only keep the error? Just making sure we're on the same page :)

File: src/__tests__/Headroom.spec.tsx [hunk 5]
```diff
@@ -31,11 +35,13 @@ describe('Headroom', () => {
     expect(component.childAt(0).props()).toEqual({
       animateUpFrom: null,
       children: <MockNode />,
-      static: true,
-      top: -50,
+      className: undefined,
+      positionStickyDisabled: false,
```
[the remark this answers is not in the document]
Comment (sarahsporck) [reply src/__tests__/Headroom.spec.tsx:41]: jap. as we **do** actually pass these values to the HeaderWrapper.

File: tools/build.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+const typescriptDeclaration = createdFiles[entryFile.replace('.tsx', '.d.ts')]
+writeFileSync('./index.d.ts', typescriptDeclaration)
+console.log('typescript declarations successfully emitted')
+
+const transpiled = transformFileSync(entryFile, {
+  minify: true
+})?.code
+if (transpiled) {
+  writeFileSync('./index.js', transpiled)
```
Comment (sarahsporck) [inline tools/build.ts:35]: I wonder if there is a simple way to test that this emitted file works properly. Especially after switching the framework I feel that we should test this.

**Review by sarahsporck — COMMENTED** [review]
I replaced babel with swc transpiler. Removes a lot of dependencies and feels much cleaner now :)

**Comment by michael-markl** [comment]
I adjusted the build script a little bit.

I added the source file now to the deployed package, so that we can have source maps and better IDE support (e.g. when going to the definition of the Headroom class, the IDE should now directly jump into the class in the source - as opposed to in the .d.ts file)

Imo this can be merged. Feel free to take another look @sarahsporck  :)

**Review by michael-markl — APPROVED** [review]

**Pull request merged**